src/blockchain/rpc.rs

```diff
@@ -178,224 +182,61 @@ impl GetBlockHash for RpcBlockchain {
-            res.into_iter()
-                .map(|v| match v["success"].as_bool() {
-                    Some(true) => Ok(()),
-                    Some(false) => Err(Error::Generic(
-                        v["error"]["message"]
-                            .as_str()
-                            .unwrap_or("Unknown error")
-                            .to_string(),
-                    )),
-                    _ => Err(Error::Generic("Unexpected response from Core".to_string())),
-                })
-                .collect::<Result<Vec<_>, _>>()?;
-        } else {
-            let requests: Vec<_> = scripts_pubkeys
-                .iter()
-                .map(|s| ImportMultiRequest {
-                    timestamp: ImportMultiRescanSince::Timestamp(0),
-                    script_pubkey: Some(ImportMultiRequestScriptPubkey::Script(s)),
-                    watchonly: Some(true),
-                    ..Default::default()
-                })
-                .collect();
-            let options = ImportMultiOptions {
-                rescan: Some(false),
+        let db_spks = db.iter_script_pubkeys(None)?;
+        let db_spk_count = db_spks.len();
+
+        // this is a hack to check whether the scripts are coming from a derivable descriptor
+        // we assume for non-derivable descriptors, the initial script count is always 1 and
+        // since we are working with two descriptors, we use 2 to determine derivability
+        let is_derivable = db_spk_count > 2;
+
+        // ensure db scripts meet start script count requirements
+        if is_derivable && db_spks.len() < self.sync_params.start_script_count {
+            let err = MissingCachedScripts {
+                last_count: db_spk_count,
+                missing_count: self.sync_params.start_script_count - db_spk_count,
             };
-            self.client.import_multi(&requests, Some(&options))?;
+            debug!("requesting more scriptPubKeys with: {:?}", err);
+            return Err(Error::MissingCachedScripts(err));
         }
 
-        loop {
-            let current_height = self.get_height()?;
-
-            // min because block invalidate may cause height to go down
-            let node_synced = self.get_node_synced_height()?.min(current_height);
-
-            let sync_up_to = node_synced.saturating_add(10_000).min(current_height);
-
-            debug!("rescan_blockchain from:{} to:{}", node_synced, sync_up_to);
-            self.client
-                .rescan_blockchain(Some(node_synced as usize), Some(sync_up_to as usize))?;
-            progress_update.update((sync_up_to as f32) / (current_height as f32), None)?;
-
-            self.set_node_synced_height(sync_up_to)?;
+        // this tells Core wallet where to sync from for imported scripts
+        let start_epoch = db
+            .get_sync_time()?
+            .map_or(self.sync_params.start_time, |st| st.block_time.timestamp);
```

Comment:
Maybe we should have a new parameter `RpcSyncParams::force_start_time` that forces every sync to use `sync_params.start_time`. This is for users that prefer reliability above all else.

Reply:
👍🏻

Reply:
Added.